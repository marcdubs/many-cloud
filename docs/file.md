# File

- [Introduction](#introduction)
- [Constructors](#constructors)
  * [new File(id, connection)](#new-fileid-connection)
- [Methods](#methods)
  * [download_to(path)](#download_topath)
  * [get_checksum()](#get_checksum)

## Introduction

File is one of the three object classes that provides abstraction to what cloud drive you're dealing with. Providing similar functions for all of them.

File and Folder both import all of the functions inside of [item](item.md).

After you establish a connection to any integration as described in the [authentication](authentication.md) documentation, you can create a file like so:
```js
const File = require("many-cloud/lib/abstractions/file");

//Create a File object with an ID and a drive connection.
let file = new File("some-file-id", connection);
```

Along with all the functions of [item](item.md), file also has these functions:

## Constructors

### new File(id, connection)

Parameters:
| Name | Type | Description |
| -- | -- | -- |
| id | String | A file ID |
| connection | Object | A connection to an integration |

## Methods

### download_to(path)

Parameters:
| Name | Type | Description |
| -- | -- | -- |
| path | String | Path to destination |

Returns: Promise that resolves once file is downloaded

### get_checksum()

Parameters: None

Returns: Promise that resolves with a checksum (sha1 or md5, etc depending on what the cloud service decided on).
