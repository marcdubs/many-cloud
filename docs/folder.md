# Folder
- [Introduction](#introduction)
- [Constructors](#constructors)
  * [new Folder(id, connection)](#new-folderid-connection)
- [Methods](#methods)
  * [list_files()](#list_files)
  * [upload_file(path)](#upload_filepath)
  * [new_folder(name)](#new_foldername)
## Introduction
Folder is one of the three object classes that provides abstraction to what cloud drive you're dealing with. Providing similar functions for all of them.

File and Folder both import all of the functions inside of [item](item.md).

After you establish a connection to any integration as described in the [authentication](authentication.md) documentation, you can create a folder like so:
```js
const Folder = require("many-cloud/lib/abstractions/folder");

//Create a File object with an ID and a drive connection.
let folder = new Folder("some-file-id", connection);
let root_folder = new Folder("root", connection);
```

Along with all the functions of [item](item.md), folder also has these functions:

## Constructors

### new Folder(id, connection)

Parameters:
| Name | Type | Description |
| -- | -- | -- |
| id | String | A Folder ID. "root" for root folder. |
| connection | Object | A connection to an integration |

## Methods

### list_files()
List the files in this folder.

Parameters: None

Returns: Promise A promise that resolves with a list of files

### upload_file(path)
Upload a local file to this folder.

Parameters:
| Name | Type | Description |
| -- | -- | -- |
| path | String | Path to local file to upload |

Returns: Promise that returns a file object when resolved

### new_folder(name)
Creates a new folder inside of this folder.

Parameters:
| Name | Type | Description |
| -- | -- | -- |
| name| String | Name of folder to create |

Returns: Promise that resolves with a Folder object.
