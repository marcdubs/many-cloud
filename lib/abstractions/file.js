const BoxFile = require("../integrations/Box/abstractions/file");
const GoogleDriveFile = require("../integrations/GoogleDrive/abstractions/file");

/**
 * File class
 * @constructor
 * @param {String} id - File id.
 * @param {Object} connection - A connection to an integration.
 */
function File(id, connection) {
  if (connection.type === "Box") {
    return new BoxFile(id, connection);
  } else if (connection.type === "GoogleDrive") {
    return new GoogleDriveFile(id, connection);
  }
}

module.exports = File;
