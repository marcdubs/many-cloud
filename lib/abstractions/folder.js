const BoxFolder = require("../integrations/Box/abstractions/folder");
const GoogleDriveFolder = require("../integrations/GoogleDrive/abstractions/folder");
const S3DriveFolder = require("../integrations/S3/abstractions/folder");

/**
 * Folder class
 * @constructor
 * @param {String} id - Folder id. "root" for root folder.
 * @param {Object} connection - A connection to an integration.
 */
function Folder(id, connection) {
  if (connection.type === "Box") {
    return new BoxFolder(id, connection);
  } else if (connection.type === "GoogleDrive") {
    return new GoogleDriveFolder(id, connection);
  } else if (connection.type === "S3") {
    return new S3DriveFolder(id, connection);
  }
}

module.exports = Folder;
