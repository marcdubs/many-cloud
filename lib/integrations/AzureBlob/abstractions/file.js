const Item = require("./item");

const fs = require("fs");
const tmp = require("tmp");
const path = require("path");
const md5File = require("md5-file");

class File extends Item {
  constructor(id, connection) {
    super(connection);

    this.type = "file";
    this.id = id;

    // Remove the first character if it is a slash
    if (this.id.charAt(0) === "/") {
      this.id = this.id.substring(1);
    }

    this.blockBlobClient = connection.containerClient.getBlockBlobClient(this.id);
  }

  async download_to(localPath) {
    if (fs.existsSync(localPath)) {
      fs.rmSync(localPath);
    }

    await this.blockBlobClient.downloadToFile(localPath);
  }

  async delete() {
    await this.blockBlobClient.delete();
  }

  /**
   * Not recommended to be heavily used for Azure, since this needs to download the file temporarily to get the checksum.
   */
  async get_checksum() {
    const tempDir = tmp.dirSync({
      unsafeCleanup: true,
    });

    let localFilePath = path.join(tempDir.name, await this.get_name());

    await this.download_to(localFilePath);

    const hash = md5File.sync(localFilePath);

    tempDir.removeCallback();

    return hash;
  }
}

module.exports = File;