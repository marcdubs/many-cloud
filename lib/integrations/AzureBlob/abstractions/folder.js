const Item = require("./item");

const path = require("path");

class Folder extends Item {
  constructor(id, connection) {
    super(connection);

    this.type = "folder";
    this.id = id;
    this.containerClient = connection.containerClient;

    if (id.toLowerCase() === "root") {
      this.id = "/";
    };

    // Remove the last character if it's a slash
    if (this.id.charAt(this.id.length - 1) === "/") {
      this.id = this.id.substring(0, this.id.length - 1);
    }

    // Remove the first character if it is a slash
    if (this.id.charAt(0) === "/") {
      this.id = this.id.substring(1);
    }
  }

  async list_files() {
    const File = require("./file");

    let list = [];

    let prefix = this.id;

    // Append a slash
    if (prefix.charAt(prefix.length - 1) !== "/") {
      prefix += "/";
    }

    // Special case for root, just make prefix empty
    if (prefix === "/") {
      prefix = "";
    }

    // Get files in this folder
    for await (const item of this.containerClient.listBlobsByHierarchy("/", { prefix })) {
      if (item.kind === "prefix") {
        list.push(new Folder(item.name, this.connection));
      } else {
        list.push(new File(item.name, this.connection));
      }
    }

    return list;
  }

  async delete() {
    const files = await this.list_files();
    for (let i = 0; i < files.length; i++) {
      await files[i].delete();
    }
  }

  async upload_file(localPath) {
    const File = require("./file");
    const fileName = path.basename(localPath);
    const file = new File(this.connection.join_path(this.id, fileName), this.connection);
    await file.blockBlobClient.uploadFile(localPath);
    return file;
  }

  async new_folder(name) {
    return new Folder(this.connection.join_path(this.id, name), this.connection);
  }
}

module.exports = Folder;