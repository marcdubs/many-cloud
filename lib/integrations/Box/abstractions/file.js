const Item = require("./item");

/**
 * File class
 * @constructor
 * @param {String} id - File id.
 * @param {Object} connection - A connection to a Box integration.
 */
function File(id, connection) {
  this.id = id;
  this.type = "file";
  this.connection = connection;
}

File.prototype = Object.create(Item.prototype);
File.prototype.constructor = File;

/**
 * Downloads this file to specified location
 * @param {String} path - Path to destination.
 *
 * @return {Promise}
 */
File.prototype.download_to = function(path) {
  return new Promise(async (resolve, reject) => {
    try {
      await this.connection.download_file(this.id, path);
      resolve();
    } catch (err) {
      reject(err);
    }
  });
};

/**
 * Gets a checksum (sha1 or md5, etc depending on implementation)
 *
 * @return {Promise} Resolves with the checksum.
 */
File.prototype.get_checksum = function() {
  return new Promise(async (resolve, reject) => {
    if (!this.checksum) {
      try {
        await this.retrieve_info();
      } catch (err) {
        reject(err);
        return;
      }
    }
    resolve(this.checksum);
  });
};

module.exports = File;
