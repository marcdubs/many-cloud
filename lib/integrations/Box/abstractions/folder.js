const Item = require("./item");

/**
 * Folder class
 * @constructor
 * @param {String} id - Folder id. "root" for root folder.
 * @param {Object} connection - A connection to a Box integration.
 */
function Folder(id, connection) {
  if (id === "root") {
    id = 0;
  }
  this.id = id;
  this.type = "folder";
  this.connection = connection;
}

Folder.prototype = Object.create(Item.prototype);
Folder.prototype.constructor = Folder;

/**
 * List the files in this folder.
 *
 * @return {Promise} A promise that returns a list of files when resolved
 */
Folder.prototype.list_files = function() {
  const File = require("./file");
  return new Promise(async (resolve, reject) => {
    try {
      let res = (await this.connection.list_all_files(this.id)).entries;
      let ret = [];
      for (let i = 0; i < res.length; i++) {
        let f;
        if (res[i].type === "folder") {
          f = new Folder(res[i].id, this.connection);
        } else {
          f = new File(res[i].id, this.connection);
        }
        f.name = res[i].name;
        ret.push(f);
      }
      resolve(ret);
    } catch (err) {
      reject(err);
    }
  });
};

/**
 * Upload a local file to this folder.
 *
 * @param path path to local file to upload
 * @return {Promise} A promise that returns a file object when resolved
 */
Folder.prototype.upload_file = function(path) {
  const File = require("./file");
  return new Promise(async (resolve, reject) => {
    try {
      let res = (await this.connection.upload_file(this.id, path)).entries[0];
      resolve(new File(res.id, this.connection));
    } catch (err) {
      reject(err);
    }
  });
};

module.exports = Folder;
