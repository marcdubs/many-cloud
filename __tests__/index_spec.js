describe("Index", () => {
  const app = require("../");
  it("app should not be null", () => {
    expect(app).not.toBeNull();
  });

  describe("integration_list", () => {
    const loaded_integration_list = require("../lib/integrations/list.json");
    it("/lib/integrations/list.json should exist", () => {
      expect(loaded_integration_list).not.toBeNull();
    });

    it("integration_list should exist", () => {
      expect(app.integration_list).not.toBeNull();
    });

    describe("all Integrations Exist", () => {
      for (let i = 0; i < loaded_integration_list.length; i++)
        it(
          'integration: "' + loaded_integration_list[i] + '" should exist',
          () => {
            expect(app.integration_list).toContainEqual(
              loaded_integration_list[i]
            );
          }
        );
    });
  });

  describe("integration", () => {
    describe("test all", () => {
      const loaded_integration_list = require("../lib/integrations/list.json");
      for (let i = 0; i < loaded_integration_list.length; i++)
        it('can get Integration: "' + loaded_integration_list[i] + '"', () => {
          expect(app.integration(loaded_integration_list[i])).not.toBeNull();
        });
    });

    describe("when argument is missing", () => {
      it("throws an error", () => {
        expect(() => {
          app.integration();
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe("when integration is unknown", () => {
      it("throws an error", () => {
        expect(() => {
          app.integration("not_an_integration");
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe("abstraction", () => {
    describe("when argument is folder", () => {
      it("returns the folder abstraction", () => {
        expect(app.abstraction("folder")).toEqual(
          require("../lib/abstractions/folder")
        );
      });
    });

    describe("when argument is file", () => {
      it("returns the file abstraction", () => {
        expect(app.abstraction("file")).toEqual(
          require("../lib/abstractions/file")
        );
      });
    });

    describe("when missing argument", () => {
      it("throws an error", () => {
        expect(() => {
          app.abstraction();
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe("when argument is not an abstraction", () => {
      it("throws an error", () => {
        expect(() => {
          app.abstraction("not_an_abstraction");
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });
});
