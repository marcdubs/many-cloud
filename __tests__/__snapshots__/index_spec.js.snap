// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index abstraction when argument is not an abstraction throws an error 1`] = `"Abstraction \\"not_an_abstraction\\" not found."`;

exports[`Index abstraction when missing argument throws an error 1`] = `"Missing abstraction argument"`;

exports[`Index integration when argument is missing throws an error 1`] = `"Missing name argument."`;

exports[`Index integration when integration is unknown throws an error 1`] = `"Integration \\"not_an_integration\\" not found."`;
